src/commands/force/mdapi/beta/convert.ts

```diff
@@ -0,0 +1,215 @@
+import { FsError } from '../../../../types';
+
+Messages.importMessagesDirectory(__dirname);
+const messages = Messages.loadMessages('@salesforce/plugin-source', 'md.convert');
+
+interface EnsureFlagOptions {
+  flagName: string;
+  path: string;
+  type: 'dir' | 'file' | 'any';
+  throwOnENOENT?: boolean;
+}
+
+export class Convert extends SourceCommand {
+  public static readonly description = messages.getMessage('description');
+  public static readonly examples = messages.getMessage('examples').split(os.EOL);
+  public static readonly requiresProject = true;
+  public static readonly flagsConfig: FlagsConfig = {
+    rootdir: flags.directory({
+      char: 'r',
+      description: messages.getMessage('flags.rootdir'),
+      longDescription: messages.getMessage('flagsLong.rootdir'),
+      required: true,
+    }),
+    outputdir: flags.directory({
+      char: 'd',
+      description: messages.getMessage('flags.outputdir'),
+      longDescription: messages.getMessage('flagsLong.outputdir'),
+    }),
+    manifest: flags.string({
+      char: 'x',
+      description: messages.getMessage('flags.manifest'),
+      longDescription: messages.getMessage('flagsLong.manifest'),
+    }),
+    metadatapath: flags.array({
+      char: 'p',
+      description: messages.getMessage('flags.metadatapath'),
+      longDescription: messages.getMessage('flagsLong.metadatapath'),
+      exclusive: ['manifest', 'metadata'],
+    }),
+    metadata: flags.array({
+      char: 'm',
+      description: messages.getMessage('flags.metadata'),
+      longDescription: messages.getMessage('flagsLong.metadata'),
+      exclusive: ['manifest', 'metadatapath'],
+    }),
+  };
+
+  private rootDir: string;
+  private outputDir: string;
+  private convertResult: ConvertResult;
+
+  public async run(): Promise<ConvertCommandResult> {
+    await this.convert();
+    this.resolveSuccess();
+    return this.formatResult();
+  }
+
+  protected async convert(): Promise<void> {
+    this.rootDir = this.resolveRootDir(this.getFlag<string>('rootdir'));
+    this.outputDir = this.resolveOutputDir(this.getFlag<string>('outputdir'));
+    const metadatapath = this.resolveMetadataPaths(this.getFlag<string[]>('metadatapath'));
+    const manifest = this.resolveManifest(this.getFlag<string>('manifest'));
+    const metadata = this.getFlag<string[]>('metadata');
+
+    let paths: string[];
+    if (metadatapath) {
+      paths = metadatapath;
+    } else if (!manifest && !metadata) {
+      paths = [this.rootDir];
+    }
+
+    this.componentSet = await ComponentSetBuilder.build({
+      sourcepath: paths,
+      manifest: manifest && {
+        manifestPath: manifest,
+        directoryPaths: [this.rootDir],
+      },
+      metadata: metadata && {
+        metadataEntries: metadata,
+        directoryPaths: [this.rootDir],
+      },
+    });
+
+    const numOfComponents = this.componentSet.getSourceComponents().toArray().length;
```

Comment:
why not this.componentSet.size?

Reply:
Because it would report a size of 1 with certain kinds of ComponentSets (built from metadata or manifest) while there were no source components.  This way is more reliable when you care about the number of source components.

Reply:
Your PRs are so educational.  Let me check my understanding because that is 100% not intuitive to me....

`.size` only gives the number of _source-backed-components_ 
but `.getSourceComponents` will give you an array that includes _non-source-backed-components_, too?